subsys/net/lib/sockets/sockets.c

```diff
@@ -1155,6 +1155,81 @@ int zsock_setsockopt(int sock, int level, int optname,
 	VTABLE_CALL(setsockopt, sock, level, optname, optval, optlen);
 }
 
+int zsock_getsockname_ctx(struct net_context *ctx, struct sockaddr *addr,
+			  socklen_t *addrlen)
+{
+	socklen_t newlen = 0;
+
+	/* If we don't have a connection handler, the socket is not bound */
+	if (ctx->conn_handler) {
+		SET_ERRNO(EINVAL);
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV4) && ctx->local.family == AF_INET) {
+		struct sockaddr_in addr4;
+
+		addr4.sin_family = AF_INET;
+		addr4.sin_port = net_sin_ptr(&ctx->local)->sin_port;
+		memcpy(&addr4.sin_addr, net_sin_ptr(&ctx->local)->sin_addr,
+		       sizeof(struct in_addr));
+		newlen = sizeof(struct sockaddr_in);
+
+		memcpy(addr, &addr4, MIN(*addrlen, newlen));
+	} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
+		   ctx->local.family == AF_INET6) {
+		struct sockaddr_in6 addr6;
+
+		addr6.sin6_family = AF_INET6;
+		addr6.sin6_port = net_sin6_ptr(&ctx->local)->sin6_port;
+		memcpy(&addr6.sin6_addr, net_sin6_ptr(&ctx->local)->sin6_addr,
+		       sizeof(struct in6_addr));
+		newlen = sizeof(struct sockaddr_in6);
+
+		memcpy(addr, &addr6, MIN(*addrlen, newlen));
+	} else {
+		SET_ERRNO(EINVAL);
+	}
+
+	*addrlen = newlen;
+
+	return 0;
+}
+
+int z_impl_zsock_getsockname(int sock, struct sockaddr *addr,
+			     socklen_t *addrlen)
+{
+	const struct fd_op_vtable *vtable;
+	void *ctx = z_get_fd_obj_and_vtable(sock, &vtable);
+
+	if (ctx == NULL) {
+		return -1;
+	}
+
+	NET_DBG("getsockname: ctx=%p, fd=%d", ctx, sock);
+
+	return z_fdtable_call_ioctl(vtable, ctx, ZFD_IOCTL_GETSOCKNAME,
+				    addr, addrlen);
+}
+
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(zsock_getsockname, sock, addr, addrlen)
+{
+	struct sockaddr addr_cpy;
+	socklen_t addrlen_cpy, addrlen_new;
+	int ret;
+
+	addrlen_cpy = addrlen_new = *((socklen_t *)addrlen);
```

Comment:
not super found of this way of writing, I'd pre-assign addrlen_cpy and addrlen_new directly at their declaration